File: 0966-binary-subarrays-with-sum/0966-binary-subarrays-with-sum.py
class Solution:
    def fuck(self, nums: List[int], goal: int) -> int:
        if goal < 0:
            return 0

        cnt, sum, l = 0, 0, 0

        for r in range(len(nums)):
            sum += nums[r]
            while sum > goal:
                sum -= nums[l]
                l += 1
            cnt += r-l+1

        return cnt 
    
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        cnt = self.fuck(nums, goal) - self.fuck(nums, goal-1)
        return cnt

File: 0966-binary-subarrays-with-sum/0966-binary-subarrays-with-sum.cpp
class Solution {
public:
    int fuck(vector<int> &nums, int goal) {
        if(goal < 0 || nums.size() == 0) return 0;

        int cnt = 0;
        int sum = 0;
        int l = 0;

        for(int r=0;r<nums.size();r++) {
            sum += nums[r];

            while(sum > goal) 
                sum -= nums[l++];
            
            cnt += r-l+1;

        }

        return cnt;
    }

    int numSubarraysWithSum(vector<int>& nums, int goal) {
        return fuck(nums, goal) - fuck(nums, goal-1);
    }
};